chain/network/src/network_protocol/mod.rs

```diff
@@ -285,23 +290,47 @@ impl PeerMessage {
     pub(crate) fn serialize(&self, enc: Encoding) -> Vec<u8> {
         match enc {
             Encoding::Borsh => borsh_::PeerMessage::from(self).try_to_vec().unwrap(),
-            Encoding::Proto => proto::PeerMessage::from(self).write_to_bytes().unwrap(),
+            Encoding::Proto => {
+                let mut msg = proto::PeerMessage::from(self);
+                let propagator = NodePropagator::new();
+                let cx = Span::current().context();
+                let mut headers = HashMap::new();
+                propagator.inject_context(&cx, &mut headers);
+                if !headers.is_empty() {
+                    msg.trace_context = NodePropagator::serialize_headers(&headers);
+                }
+                msg.write_to_bytes().unwrap()
+            }
         }
     }
 
     pub(crate) fn deserialize(
         enc: Encoding,
         data: &[u8],
     ) -> Result<PeerMessage, ParsePeerMessageError> {
+        let span = tracing::trace_span!(target: "network", "deserialize").entered();
         Ok(match enc {
             Encoding::Borsh => (&borsh_::PeerMessage::try_from_slice(data)
                 .map_err(ParsePeerMessageError::BorshDecode)?)
                 .try_into()
                 .map_err(ParsePeerMessageError::BorshConv)?,
-            Encoding::Proto => (&proto::PeerMessage::parse_from_bytes(data)
-                .map_err(ParsePeerMessageError::ProtoDecode)?)
-                .try_into()
-                .map_err(ParsePeerMessageError::ProtoConv)?,
+            Encoding::Proto => {
+                let proto_msg: proto::PeerMessage = proto::PeerMessage::parse_from_bytes(data)
+                    .map_err(ParsePeerMessageError::ProtoDecode)?;
+                if !proto_msg.trace_context.is_empty() {
+                    let propagator = NodePropagator::new();
+                    let headers = NodePropagator::deserialize_headers(&proto_msg.trace_context);
+                    if let Ok(headers) = headers {
+                        let extracted_span_context = propagator.extract_span_context(&headers);
+                        span.clone()
+                            .or_current()
+                            .add_link(extracted_span_context.as_ref().unwrap().clone());
+                    } else {
+                        tracing::warn!("Failed to parse headers: {:#?}", headers);
```

Comment:
This seems like a risky warning to have. I imagine that if a node fails to parse headers once, it will likely continue failing for most of the messages, which likely means there is gonna be a _lot_ of log spam when a node is in such a situation. I’d argue that we should either drop down the level significantly here, or perhaps only output every, say 1000th message, or some combination of both.

Reply:
Removed